cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java

```diff
@@ -0,0 +1,129 @@
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.NamespaceId;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Inject;
+import org.apache.tephra.RetryStrategies;
+import org.apache.tephra.TransactionExecutorFactory;
+import org.apache.tephra.TransactionFailureException;
+import org.apache.tephra.TransactionSystemClient;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Default implementation of the {@link PreviewStore} that stores data in in-memory table
+ */
+public class DefaultPreviewStore implements PreviewStore {
+
+  private static final DatasetId PREVIEW_TABLE_ID = NamespaceId.SYSTEM.dataset(PreviewDataset.PREVIEW_TABLE_NAME);
+
+  private final DatasetFramework dsFramework;
+  private final Transactional transactional;
+
+  @Inject
+  public DefaultPreviewStore(DatasetFramework framework, TransactionSystemClient txClient) {
+    this.dsFramework = framework;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(
+        new SystemDatasetInstantiator(framework), txClient,
+        NamespaceId.SYSTEM, ImmutableMap.<String, String>of(), null, null)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+  }
+
+  @Override
+  public void put(final ApplicationId applicationId, final String tracerName, final String propertyName,
+                  final Object value) {
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          getPreviewDataset(context).put(applicationId, tracerName, propertyName, value);
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e);
+    }
+  }
+
+  @Override
+  public Map<String, List<String>> get(final ApplicationId applicationId, final String tracerName) {
+    try {
+      return Transactions.execute(transactional, new TxCallable<Map<String, List<String>>>() {
+        @Override
+        public Map<String, List<String>> call(DatasetContext context) throws Exception {
+          return getPreviewDataset(context).get(applicationId, tracerName);
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e);
+    }
+  }
+
+  @Override
+  public void remove(final ApplicationId applicationId) {
+    try {
+      transactional.execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          getPreviewDataset(context).remove(applicationId);
+        }
+      });
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e);
+    }
+  }
+
+  @VisibleForTesting
+  @Override
+  public void clear() throws IOException, DatasetManagementException {
```

Comment:
For method that is visible for testing, it shouldn't be `public`, but instead should be package protected. Also, I don't see any unit-test in this PR, how do you know this method is needed to be visible for testing?